File: bridge/telemetry.py
"""Hermes Bridge - telemetry center.

In-process signal bus + sliding-window request log. Lets every endpoint /
module emit structured events that the signal panel can later aggregate.

Two layers:

1. SignalBus - pub/sub, in-memory, no persistence. Subscribers register a
   callback for a topic; producers call emit(topic, payload). The bus is
   fire-and-forget by design (signals are best-effort, never block the
   request path).

2. RequestLog - ring buffer (default 1000 entries) of structured access
   records. Each entry captures method, path, status, elapsed_ms, payload
   sizes, error (if any). Thread-safe (threading.Lock); designed for both
   async and sync callers.

Persists across requests but does NOT persist across restarts (signals
are ephemeral - restart is itself a "first event").
"""
from __future__ import annotations

import json
import os
import threading
import time
import uuid
from collections import deque
from pathlib import Path
from typing import Callable, Deque, Dict, List, Optional


def _telemetry_path() -> Path:
    base = Path(os.environ.get("HERMES_LOGS") or
                (Path(__file__).resolve().parent.parent / "data" / "logs"))
    base.mkdir(parents=True, exist_ok=True)
    return base / "telemetry.json"


class SignalBus:
    """Topic-based pub/sub for in-process signals.

    Subscribers register fn(topic, payload) -> None for a topic string.
    Producers call emit(topic, payload). Wildcards: subscribe to "*" to
    receive everything.

    Concurrency: subscribers are called synchronously in emit(). Failures
    in one subscriber do NOT affect others - they are isolated.
    """

    def __init__(self, max_history: int = 200) -> None:
        self._subs: Dict[str, List[Callable[[str, dict], None]]] = {}
        self._lock = threading.Lock()
        self._history: Deque[dict] = deque(maxlen=max_history)

    def subscribe(self, topic: str, fn: Callable[[str, dict], None]) -> Callable[[], None]:
        with self._lock:
            self._subs.setdefault(topic, []).append(fn)

        def _unsub() -> None:
            with self._lock:
                if topic in self._subs and fn in self._subs[topic]:
                    self._subs[topic].remove(fn)

        return _unsub

    def emit(self, topic: str, payload: Optional[dict] = None) -> dict:
        """Fire a signal. Returns the envelope that was broadcast."""
        envelope: dict = {
            "id": uuid.uuid4().hex[:16],
            "ts": time.time(),
            "topic": topic,
            "payload": payload or {},
            "source": os.environ.get("HERMES_MODULE", "bridge"),
        }
        with self._lock:
            self._history.append(envelope)
            subs = list(self._subs.get(topic, [])) + list(self._subs.get("*", []))
        for fn in subs:
            try:
                fn(topic, envelope)
            except Exception:
                pass
        return envelope

    def recent(self, topic: Optional[str] = None, limit: int = 50) -> List[dict]:
        with self._lock:
            items = list(self._history)
        if topic:
            items = [e for e in items if e["topic"] == topic or e["topic"].startswith(topic + ".")]
        return items[-limit:]

class RequestLog:
    """Ring-buffer access log with per-path aggregates."""

    def __init__(self, capacity: int = 1000) -> None:
        self._buf: Deque[dict] = deque(maxlen=capacity)
        self._lock = threading.Lock()
        self._started = time.time()

    def record(self, entry: dict) -> None:
        entry.setdefault("id", f"req-{uuid.uuid4().hex[:12]}")
        entry.setdefault("ts", time.time())
        entry.setdefault("module", os.environ.get("HERMES_MODULE", "bridge"))
        with self._lock:
            self._buf.append(entry)

    def recent(self, limit: int = 100) -> List[dict]:
        with self._lock:
            return list(self._buf)[-limit:]

    def stats(self) -> dict:
        with self._lock:
            entries = list(self._buf)
        if not entries:
            return {
                "uptime_sec": round(time.time() - self._started, 1),
                "total": 0, "errors": 0, "by_path": {}, "by_status": {},
            }

        total = len(entries)
        errors = sum(1 for e in entries if (e.get("status", 0) or 0) >= 500 or e.get("error"))
        by_status: Dict[str, int] = {}
        by_path_raw: Dict[str, List[float]] = {}
        for e in entries:
            st = e.get("status")
            if st is not None:
                by_status[str(st)] = by_status.get(str(st), 0) + 1
            p = e.get("path", "?")
            by_path_raw.setdefault(p, []).append(float(e.get("elapsed_ms", 0) or 0))

        def _percentile(xs: List[float], pct: float) -> float:
            if not xs:
                return 0.0
            xs = sorted(xs)
            i = max(0, min(len(xs) - 1, int(len(xs) * pct / 100)))
            return round(xs[i], 2)

        by_path: Dict[str, dict] = {}
        for p, xs in by_path_raw.items():
            by_path[p] = {
                "count": len(xs),
                "errors": sum(1 for e in entries if e.get("path") == p and ((e.get("status") or 0) >= 500 or e.get("error"))),
                "p50_ms": _percentile(xs, 50),
                "p95_ms": _percentile(xs, 95),
                "max_ms": round(max(xs), 2),
            }

        return {
            "uptime_sec": round(time.time() - self._started, 1),
            "total": total,
            "errors": errors,
            "error_rate": round(errors / total, 4),
            "by_status": dict(sorted(by_status.items())),
            "by_path": dict(sorted(by_path.items())),
        }

    def flush(self, path: Optional[Path] = None) -> None:
        target = path or _telemetry_path()
        snapshot = {
            "ts": time.time(),
            "stats": self.stats(),
            "recent": self.recent(limit=100),
        }
        tmp = target.with_suffix(target.suffix + ".tmp")
        try:
            tmp.write_text(json.dumps(snapshot, ensure_ascii=False, indent=1), encoding="utf-8")
            os.replace(tmp, target)
        except OSError:
            if tmp.exists():
                try:
                    tmp.unlink()
                except OSError:
                    pass


# ---- Module-level singletons ----

_bus = SignalBus(max_history=int(os.environ.get("HERMES_TELEMETRY_HISTORY", "200")))
_log = RequestLog(capacity=int(os.environ.get("HERMES_TELEMETRY_CAPACITY", "1000")))


def bus() -> SignalBus:
    return _bus


def log() -> RequestLog:
    return _log


def reset_for_tests() -> None:
    global _bus, _log
    _bus = SignalBus(max_history=200)
    _log = RequestLog(capacity=1000)


class Topics:
    """Canonical signal topics - use these constants to avoid drift."""

    MODULE_BOOT = "module.boot"
    MODULE_READY = "module.ready"
    MODULE_SHUTDOWN = "module.shutdown"
    MODULE_ERROR = "module.error"
    MODEL_LOADED = "model.loaded"
    MODEL_EVICTED = "model.evicted"
    MODEL_SWAP = "model.swap"
    MODEL_WARMUP_START = "model.warmup.start"
    MODEL_WARMUP_PROGRESS = "model.warmup.progress"
    MODEL_WARMUP_DONE = "model.warmup.done"
    CHAT_REQUEST = "chat.request"
    CHAT_DELTA = "chat.delta"
    CHAT_DONE = "chat.done"
    CHAT_ERROR = "chat.error"
    SESSION_OPENED = "session.opened"
    SESSION_CLOSED = "session.closed"
    SUPERVISOR_HEARTBEAT = "supervisor.heartbeat"
    PORT_LISTEN = "port.listen"
    PORT_LOST = "port.lost"
    DISK_LOW = "disk.low"


__all__ = [
    "SignalBus", "RequestLog", "Topics",
    "bus", "log", "reset_for_tests",
]

File: bridge-rs/src/telemetry.rs
//! telemetry.rs — in-process signal bus + ring-buffer request log.
//!
//! Extracted from main.rs (2026-06-28) to keep main.rs focused on routing.
//! Mirrors the deleted bridge/telemetry.py (SignalBus + RequestLog + Topics).

use serde::Serialize;
use std::collections::HashMap;
use std::sync::atomic::{AtomicU64, Ordering};
use std::time::Instant;
use tokio::sync::RwLock;

// ---------------------------------------------------------------------------
// Signal Bus
// ---------------------------------------------------------------------------

#[derive(Clone, Debug, Serialize)]
pub struct SignalEnvelope {
    pub id: u64,
    pub topic: String,
    pub payload: serde_json::Value,
    pub ts: f64,
}

pub struct SignalBus {
    envelopes: RwLock<Vec<SignalEnvelope>>,
    next_id: AtomicU64,
    capacity: usize,
}

impl SignalBus {
    pub fn new(capacity: usize) -> Self {
        Self {
            envelopes: RwLock::new(Vec::with_capacity(capacity)),
            next_id: AtomicU64::new(1),
            capacity,
        }
    }

    pub async fn emit(&self, topic: &str, payload: serde_json::Value) -> SignalEnvelope {
        let id = self.next_id.fetch_add(1, Ordering::Relaxed);
        let env = SignalEnvelope {
            id,
            topic: topic.to_string(),
            payload,
            ts: now_unix(),
        };
        let mut buf = self.envelopes.write().await;
        buf.push(env.clone());
        if buf.len() > self.capacity {
            let excess = buf.len() - self.capacity;
            buf.drain(0..excess);
        }
        env
    }

    pub async fn recent(&self, topic: Option<&str>, limit: usize) -> Vec<SignalEnvelope> {
        let buf = self.envelopes.read().await;
        let mut items: Vec<SignalEnvelope> = buf
            .iter()
            .rev()
            .filter(|e| topic.map_or(true, |t| e.topic.starts_with(t)))
            .take(limit)
            .cloned()
            .collect();
        items.reverse();
        items
    }
}

// ---------------------------------------------------------------------------
// Request Log (ring buffer)
// ---------------------------------------------------------------------------

pub struct RequestLog {
    entries: RwLock<Vec<RequestEntry>>,
    start_time: Instant,
    capacity: usize,
}

#[derive(Clone, Debug, Serialize)]
pub struct RequestEntry {
    pub path: String,
    pub status: u16,
    pub elapsed_ms: u64,
    pub ts: f64,
}

impl RequestLog {
    pub fn new(capacity: usize) -> Self {
        Self {
            entries: RwLock::new(Vec::with_capacity(capacity)),
            start_time: Instant::now(),
            capacity,
        }
    }

    pub async fn record(&self, path: &str, status: u16, elapsed_ms: u64) {
        let entry = RequestEntry {
            path: path.to_string(),
            status,
            elapsed_ms,
            ts: now_unix(),
        };
        let mut buf = self.entries.write().await;
        buf.push(entry);
        if buf.len() > self.capacity {
            let excess = buf.len() - self.capacity;
            buf.drain(0..excess);
        }
    }

    pub async fn stats(&self) -> serde_json::Value {
        let buf = self.entries.read().await;
        let total = buf.len();
        let errors = buf.iter().filter(|e| e.status >= 400).count();
        let mut by_status: HashMap<String, usize> = HashMap::new();
        let mut by_path: HashMap<String, usize> = HashMap::new();
        for e in buf.iter() {
            *by_status.entry(e.status.to_string()).or_insert(0) += 1;
            *by_path.entry(e.path.clone()).or_insert(0) += 1;
        }
        serde_json::json!({
            "total": total,
            "errors": errors,
            "error_rate": if total > 0 { errors as f64 / total as f64 } else { 0.0 },
            "uptime_sec": self.start_time.elapsed().as_secs(),
            "by_status": by_status,
            "by_path": by_path,
        })
    }

    /// Flush telemetry snapshot to data/logs/telemetry.json (non-blocking).
    pub async fn flush(&self) {
        let snapshot = serde_json::json!({
            "ts": now_unix(),
            "stats": self.stats().await,
            "recent": self.recent_entries(100).await,
        });
        // Non-blocking write
        tokio::spawn(async move {
            let log_dir = std::env::var("HERMES_LOGS")
                .map(std::path::PathBuf::from)
                .unwrap_or_else(|_| {
                    let root = std::env::current_dir().unwrap_or_default();
                    root.join("data").join("logs")
                });
            let _ = std::fs::create_dir_all(&log_dir);
            let path = log_dir.join("telemetry.json");
            let tmp = path.with_extension("json.tmp");
            if let Ok(json_str) = serde_json::to_string_pretty(&snapshot) {
                if std::fs::write(&tmp, json_str).is_ok() {
                    let _ = std::fs::rename(&tmp, &path);
                }
            }
        });
    }

    async fn recent_entries(&self, limit: usize) -> Vec<RequestEntry> {
        let buf = self.entries.read().await;
        buf.iter().rev().take(limit).cloned().collect::<Vec<_>>()
            .into_iter()
            .rev()
            .collect()
    }
}

// ---------------------------------------------------------------------------
// Topics — canonical signal topic constants (mirrors bridge/telemetry.py Topics)
// ---------------------------------------------------------------------------

pub struct Topics;

impl Topics {
    // Module lifecycle
    pub const MODULE_BOOT: &'static str = "module.boot";
    pub const MODULE_READY: &'static str = "module.ready";
    pub const MODULE_SHUTDOWN: &'static str = "module.shutdown";
    pub const MODULE_ERROR: &'static str = "module.error";

    // Model lifecycle
    pub const MODEL_LOADED: &'static str = "model.loaded";
    pub const MODEL_EVICTED: &'static str = "model.evicted";
    pub const MODEL_SWAP: &'static str = "model.swap";
    pub const MODEL_WARMUP_START: &'static str = "model.warmup.start";
    pub const MODEL_WARMUP_PROGRESS: &'static str = "model.warmup.progress";
    pub const MODEL_WARMUP_DONE: &'static str = "model.warmup.done";

    // Chat
    pub const CHAT_REQUEST: &'static str = "chat.request";
    pub const CHAT_DELTA: &'static str = "chat.delta";
    pub const CHAT_DONE: &'static str = "chat.done";
    pub const CHAT_ERROR: &'static str = "chat.error";

    // Session
    pub const SESSION_OPENED: &'static str = "session.opened";
    pub const SESSION_CLOSED: &'static str = "session.closed";

    // Infrastructure
    pub const SUPERVISOR_HEARTBEAT: &'static str = "supervisor.heartbeat";
    pub const PORT_LISTEN: &'static str = "port.listen";
    pub const PORT_LOST: &'static str = "port.lost";
    pub const DISK_LOW: &'static str = "disk.low";

    // Liveness
    pub const LIVENESS_OK: &'static str = "liveness.ok";
    pub const LIVENESS_DEGRADED: &'static str = "liveness.degraded";
    pub const LIVENESS_DEAD: &'static str = "liveness.dead";
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

fn now_unix() -> f64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs_f64()
}
